File: src/main/java/ru/mydesignstudio/database/metadata/extractor/extractors/pk/PrimaryKeyExtractor.java
package ru.mydesignstudio.database.metadata.extractor.extractors.pk;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import ru.mydesignstudio.database.metadata.extractor.extractors.ExtractHelper;
import ru.mydesignstudio.database.metadata.extractor.extractors.TableMetadataAppender;
import ru.mydesignstudio.database.metadata.extractor.extractors.model.TableMetadata;
import ru.mydesignstudio.database.metadata.extractor.resource.StringResource;

@Component
public class PrimaryKeyExtractor implements TableMetadataAppender {
  @StringResource("classpath:sql/extract_primary_keys.sql")
  private String extractQuery;

  @Autowired
  private ExtractHelper helper;

  @Override
  public void append(TableMetadata metadata, String schemaName, String tableName) {
    metadata.setPrimaryKeys(helper.extract(extractQuery, new Object[] { tableName, schemaName }, PrimaryKeyModel.class));
  }
}

File: extractor-source/source-oracle/src/main/kotlin/ru/mydesignstudio/database/metadata/extractor/table/extractor/PrimaryKeyExtractor.kt
package ru.mydesignstudio.database.metadata.extractor.table.extractor

import org.springframework.stereotype.Component
import ru.mydesignstudio.database.metadata.extractor.ExtractorBase
import ru.mydesignstudio.database.metadata.extractor.connection.DataSourceProvider
import ru.mydesignstudio.database.metadata.extractor.extract.helper.QueryProvider
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataObject
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataSchema
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataSourceConnection
import ru.mydesignstudio.database.metadata.extractor.extract.result.PrimaryKeyModel

@Component
class PrimaryKeyExtractor (
        private val dataSourceProvider: DataSourceProvider,
        private val queryProvider: QueryProvider) : ExtractorBase() {
    fun extract(metadata: MetadataObject, schema: MetadataSchema, connection: MetadataSourceConnection): Collection<PrimaryKeyModel> {
        val query = queryProvider.provide("classpath:sql/extract_primary_keys.sql")
        val dataSource = dataSourceProvider.provide(connection)
        return extract(dataSource, query, listOf(metadata.name, schema.name)) { rs ->
                    PrimaryKeyModel(
                            primaryKey = getString(rs, "primary_key"),
                            schemaName = getString(rs, "schema_name"),
                            constraintName = getString(rs, "constraint_name"),
                            tableName = getString(rs, "table_name"),
                            columns = getString(rs, "columns")
                    )
                }
    }
}